src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp

```diff
@@ -0,0 +1,161 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandler.h>
+#include <app/MessageDef/AttributeDataElement.h>
+#include <app/util/af.h>
+#include <app/util/attribute-storage.h>
+#include <lib/core/Optional.h>
+#include <platform/PlatformManager.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::GeneralDiagnostics::Attributes;
+using chip::DeviceLayer::PlatformManager;
+
+namespace {
+
+class GeneralDiagosticsAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the GeneralDiagnostics cluster on all endpoints.
+    GeneralDiagosticsAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), GeneralDiagnostics::Id) {}
+
+    CHIP_ERROR Read(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter, bool * aDataRead) override;
+
+private:
+    CHIP_ERROR ReadUint8IfSupported(CHIP_ERROR (PlatformManager::*getter)(uint8_t &), TLV::TLVWriter * aWriter);
+    CHIP_ERROR ReadUint16IfSupported(CHIP_ERROR (PlatformManager::*getter)(uint16_t &), TLV::TLVWriter * aWriter);
+    CHIP_ERROR ReadUint32IfSupported(CHIP_ERROR (PlatformManager::*getter)(uint32_t &), TLV::TLVWriter * aWriter);
+    CHIP_ERROR ReadUint64IfSupported(CHIP_ERROR (PlatformManager::*getter)(uint64_t &), TLV::TLVWriter * aWriter);
+};
+
+GeneralDiagosticsAttrAccess gAttrAccess;
+
+CHIP_ERROR GeneralDiagosticsAttrAccess::Read(ClusterInfo & aClusterInfo, TLV::TLVWriter * aWriter, bool * aDataRead)
+{
+    if (aClusterInfo.mClusterId != GeneralDiagnostics::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    *aDataRead = true;
+    switch (aClusterInfo.mFieldId)
+    {
+    case Ids::RebootCount: {
+        return ReadUint16IfSupported(&PlatformManager::GetRebootCount, aWriter);
+    }
+    case Ids::UpTime: {
+        return ReadUint64IfSupported(&PlatformManager::GetUpTime, aWriter);
+    }
+    case Ids::TotalOperationalHours: {
+        return ReadUint32IfSupported(&PlatformManager::GetTotalOperationalHours, aWriter);
+    }
+    case Ids::BootReasons: {
+        return ReadUint8IfSupported(&PlatformManager::GetBootReasons, aWriter);
+    }
+    default: {
+        *aDataRead = false;
+        break;
+    }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GeneralDiagosticsAttrAccess::ReadUint8IfSupported(CHIP_ERROR (PlatformManager::*getter)(uint8_t &),
```

Comment:
These "IfSupported" methods don't do much to reduce code duplication, since each has only one caller...  Might be clearer to define a template to reduce the copy/paste or something...  It would still be just as much generated code, but at least less manual editing.

Reply:
Acked, that is a good idea